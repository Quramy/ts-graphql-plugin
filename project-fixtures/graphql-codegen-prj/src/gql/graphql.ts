/* eslint-disable */
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
};

export type Post = {
  __typename?: 'Post';
  author: User;
  body: Scalars['String']['output'];
  createdAt: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  title: Scalars['String']['output'];
  updatedAt: Scalars['String']['output'];
};

export type Query = {
  __typename?: 'Query';
  popularPosts: Array<Post>;
};

export type User = {
  __typename?: 'User';
  avatarURL: Scalars['String']['output'];
  createdAt: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  name: Scalars['String']['output'];
  updatedAt: Scalars['String']['output'];
};

export type PopularPosts_QueryQueryVariables = Exact<{ [key: string]: never; }>;


export type PopularPosts_QueryQuery = { __typename?: 'Query', popularPosts: Array<(
    { __typename?: 'Post', id: string }
    & { ' $fragmentRefs'?: { 'PostSummary_PostFragment': PostSummary_PostFragment } }
  )> };

export type PostSummary_PostFragment = { __typename?: 'Post', id: string, title: string, author: (
    { __typename?: 'User', name: string }
    & { ' $fragmentRefs'?: { 'UserAvatar_UserFragment': UserAvatar_UserFragment } }
  ) } & { ' $fragmentName'?: 'PostSummary_PostFragment' };

export type UserAvatar_UserFragment = { __typename?: 'User', name: string, avatarURL: string } & { ' $fragmentName'?: 'UserAvatar_UserFragment' };

export const UserAvatar_UserFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}}]} as unknown as DocumentNode<UserAvatar_UserFragment, unknown>;
export const PostSummary_PostFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"PostSummary_Post"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Post"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"author"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"UserAvatar_User"}}]}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}}]} as unknown as DocumentNode<PostSummary_PostFragment, unknown>;
export const PopularPosts_QueryDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"PopularPosts_Query"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"popularPosts"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"PostSummary_Post"}}]}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"PostSummary_Post"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Post"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"author"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"UserAvatar_User"}}]}}]}}]} as unknown as DocumentNode<PopularPosts_QueryQuery, PopularPosts_QueryQueryVariables>;