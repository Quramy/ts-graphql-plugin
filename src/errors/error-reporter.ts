import path from 'path';
import { TsGqlError, ErrorWithLocation, ErrorWithoutLocation } from '.';
import { pos2location, pad, color } from '../string-util';

const lineMark = (line: number, width: number) => {
  const strLine = line + 1 + '';
  return color.invert(pad(' ', width - strLine.length) + strLine) + ' ';
};

const lineMarkForUnderline = (width: number) => {
  return color.invert(pad(' ', width)) + ' ';
};

export class ErrorReporter {
  constructor(private readonly _currentDirectory: string, private readonly _output: (msg: string) => void = () => {}) {}

  outputError(error: TsGqlError) {
    if (error instanceof ErrorWithoutLocation) {
      this._outputErrorWithoutLocation(error);
    } else if (error instanceof ErrorWithLocation) {
      this._outputErrorWithLocation(error);
    }
  }

  _outputErrorWithoutLocation(error: ErrorWithoutLocation) {
    const prefix = error.severity === 'Error' ? color.red('error') : color.yellow('warn');
    this._output(prefix + ': ' + error.message);
  }

  _outputErrorWithLocation(error: ErrorWithLocation) {
    const {
      message,
      errorContent: { content, start, end, fileName },
    } = error;
    const startLC = pos2location(content, start);
    const endLC = pos2location(content, end);
    const relativeContentPath = path.isAbsolute(fileName) ? path.relative(this._currentDirectory, fileName) : fileName;
    const fileIndicator = `${relativeContentPath}:${startLC.line + 1}:${startLC.character + 1}`;
    const outputs = [`${color.thin(fileIndicator)} - ${message}`, ''];
    const allLines = content.split('\n');
    const preLines = allLines.slice(Math.max(startLC.line - 1, 0), startLC.line);
    const lines = allLines.slice(startLC.line, endLC.line + 1);
    const postLines = allLines.slice(endLC.line + 1, Math.min(allLines.length - 1, endLC.line + 2));
    const lineMarkerWidth = (Math.min(allLines.length - 1, endLC.line + 2) + '').length;
    for (let i = 0; i < preLines.length; ++i) {
      outputs.push(lineMark(i + startLC.line - 1, lineMarkerWidth) + color.thin(preLines[i]));
    }
    for (let i = 0; i < lines.length; ++i) {
      outputs.push(lineMark(i + startLC.line, lineMarkerWidth) + lines[i]);
      if (i === 0) {
        if (startLC.line === endLC.line) {
          outputs.push(
            lineMarkForUnderline(lineMarkerWidth) +
              pad(' ', startLC.character) +
              color.red(pad('~', endLC.character - startLC.character)),
          );
        } else {
          outputs.push(
            lineMarkForUnderline(lineMarkerWidth) +
              pad(' ', startLC.character) +
              color.red(pad('~', lines[i].length - startLC.character)),
          );
        }
      } else if (i === lines.length - 1) {
        outputs.push(lineMarkForUnderline(lineMarkerWidth) + color.red(pad('~', endLC.character)));
      } else {
        outputs.push(lineMarkForUnderline(lineMarkerWidth) + color.red(pad('~', lines[i].length)));
      }
    }
    for (let i = 0; i < postLines.length; ++i) {
      outputs.push(lineMark(i + endLC.line + 1, lineMarkerWidth) + color.thin(postLines[i]));
    }
    outputs.push('');
    const result = outputs.join('\n');
    this._output(result);
    return result;
  }
}
