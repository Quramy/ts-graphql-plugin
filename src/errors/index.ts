export { ErrorReporter } from './error-reporter';

export type ErrorRange = {
  fileName: string;
  start: number;
  end: number;
};

export type Severity = 'Error' | 'Warn';

export type ErrorContent = ErrorRange & {
  severity?: Severity;
  content: string;
};

export class ErrorWithLocation extends Error {
  readonly name = 'ErrorWithLocation';
  readonly severity: Severity = 'Error';

  constructor(public readonly message: string, public readonly errorContent: ErrorContent) {
    super(message);
    if (errorContent.severity) {
      this.severity = errorContent.severity;
    }
  }
}

export class ErrorWithoutLocation extends Error {
  readonly name = 'ErrorWithoutLocation';
  constructor(public readonly message: string, public readonly severity: Severity = 'Error') {
    super(message);
  }
}

export type TsGqlError = ErrorWithLocation | ErrorWithoutLocation;

export const ERROR_CODES = {
  graphqlLangServiceError: {
    code: 51001,
  },
  templateIsTooComplex: {
    code: 51010,
    message: 'This operation or fragment has too complex interpolation to analyze.',
  },
  errorInOtherInterpolation: {
    code: 51011,
    message: 'This expression has some GraphQL errors.',
  },
  schemaBuildError: {
    code: 51020,
  },
};
