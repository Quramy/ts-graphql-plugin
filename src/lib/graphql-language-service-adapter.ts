import * as ts from 'typescript/lib/tsserverlibrary';

import { buildClientSchema } from 'graphql';
import { CompletionItem, getAutocompleteSuggestions } from 'graphql-language-service-interface';
import { isTagged, TagCondition } from './ts-util/index';

export interface GraphQLLanguageServiceAdapterCreateOptions {
  schema?: any;
  logger?: (msg: string) => void;
  tag?: string;
}

export type GetCompletionAtPosition = ts.LanguageService['getCompletionsAtPosition'];

export class GraphQLLanguageServiceAdapter {

  private _schema: any;
  private _tagCondition: TagCondition = null;

  constructor(
    private _getNode: (fileName: string, position) => ts.Node = () => null,
    opt: GraphQLLanguageServiceAdapterCreateOptions = { },
  ) {
      if (opt.logger) this._logger = opt.logger;
      if (opt.schema) this.updateSchema(opt.schema);
      if (opt.tag) this._tagCondition = opt.tag;
  }

  updateSchema(schema: { data: any }) {
    try {
      this._schema = buildClientSchema(schema.data);
    } catch (err) {
      this._logger('Fail to build schema...');
      this._logger(err);
      this._schema = null;
    }
  }

  getCompletionAtPosition(delegate: GetCompletionAtPosition, fileName: string, position: number ) {
    if (!this._schema) return delegate(fileName, position);
    const node = this._getNode(fileName, position);
    if (!node || node.kind !== ts.SyntaxKind.NoSubstitutionTemplateLiteral) {
      return delegate(fileName, position);
    }
    if (this._tagCondition && !isTagged(node, this._tagCondition)) {
      return delegate(fileName, position);
    }
    const cursor = position - node.getStart();
    const text = node.getText().slice(1, cursor + 1);  // remove the backquote char
    this._logger('Search text: "' + text + '"');
    const gqlCompletionItems = getAutocompleteSuggestions(this._schema, text, cursor);
    this._logger(JSON.stringify(gqlCompletionItems));
    return translateCompletionItems(gqlCompletionItems);
  }

  private _logger: (msg: string) => void = () => { };
}

function translateCompletionItems(items: CompletionItem[]): ts.CompletionInfo {
  const result: ts.CompletionInfo = {
    isGlobalCompletion: false,
    isMemberCompletion: false,
    isNewIdentifierLocation: false,
    entries: items.map(r => {
      const kind = r.kind ? r.kind + '' : 'unknown';
      return {
        name: r.label,
        kindModifiers: 'declare',
        kind,
        sortText: '0',
      };
    }),
  };
  return result;
}
