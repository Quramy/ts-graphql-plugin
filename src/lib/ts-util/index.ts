import * as ts from 'typescript/lib/tsserverlibrary';

export type TagCondition = string;

export function findNode(sourceFile: ts.SourceFile, position: number): ts.Node | undefined {
  function find(node: ts.Node): ts.Node|undefined {
    if (position >= node.getStart() && position < node.getEnd()) {
      return ts.forEachChild(node, find) || node;
    }
  }
  return find(sourceFile);
}

export function isTagged(node: ts.Node, condition: TagCondition) {
  if (!node || !node.parent) return false;
  if (node.parent.kind !== ts.SyntaxKind.TaggedTemplateExpression) return false;
  const tagNode = node.parent as ts.TaggedTemplateExpression;
  return tagNode.tag.getText() === condition;
}
