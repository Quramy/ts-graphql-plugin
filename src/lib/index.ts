import * as ts from 'typescript/lib/tsserverlibrary';
import { GraphQLLanguageServiceAdapter } from './graphql-language-service-adapter';
import { LanguageServiceProxyBuilder } from './language-service-proxy-builder';
import { SchamaJsonManager } from './schema-json-manager';
import { findNode } from './ts-util';

function create(info: ts.server.PluginCreateInfo): ts.LanguageService {
  const getNode = (fileName: string, position: number) => {
    return findNode(info.languageService.getProgram().getSourceFile(fileName), position);
  };
  const logger = (msg: string) => info.project.projectService.logger.info(msg);
  const schemaManager = new SchamaJsonManager(info);
  const schema = schemaManager.getSchema();
  const adapter = new GraphQLLanguageServiceAdapter(getNode, { schema, logger });

  const proxy = new LanguageServiceProxyBuilder(info)
    .wrap('getCompletionsAtPosition', delegate => adapter.getCompletionPosition.bind(adapter, delegate))
    .build()
  ;

  schemaManager.registerOnChange(adapter.updateSchema.bind(adapter));
  schemaManager.startWatch();

  return proxy;
}

export const moduleFactory: ts.server.PluginModuleFactory = (mod: { typescript: typeof ts }) => {
  return { create };
};
