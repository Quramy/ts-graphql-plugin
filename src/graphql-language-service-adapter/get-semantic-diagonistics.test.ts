import * as ts from 'typescript/lib/tsserverlibrary';
import { AdapterFixture } from './testing/adapter-fixture';
import { createSimpleSchema } from './testing/simple-schema';
import { GraphQLSchema } from 'graphql';
import { mark, Frets } from 'fretted-strings';
import { ERROR_CODES } from '../errors';

function craeteFixture(name: string, schema?: GraphQLSchema) {
  return new AdapterFixture(name, schema);
}

const defaultSemanticDigostics: ts.Diagnostic[] = [];

const delegateFn = () => defaultSemanticDigostics;

describe('getSemanticDiagnostics', () => {
  it('should delegate original fn when schema is not set', () => {
    const fixture = craeteFixture('input.ts');
    const actual = fixture.adapter.getSemanticDiagnostics(delegateFn, 'input.ts');
    expect(actual).toEqual(defaultSemanticDigostics);
  });

  it('should report errors coused by schema building', () => {
    const fixture = craeteFixture('input.ts');
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');
    fixture.adapter.updateSchema(
      [
        {
          fileName: 'schema.graphql',
          message: 'some schema build error',
          fileContent: '',
          locations: [{ line: 0, character: 0 }],
        },
      ],
      null,
    );
    fixture.source = 'const query = `query { hello }`;';
    const actual = validateFn();
    expect(actual[0].code).toBe(ERROR_CODES.schemaBuildError.code);
    expect(actual[0].messageText).toMatchSnapshot();
  });

  it('should return empty array for valid GraphQL template string', () => {
    const fixture = craeteFixture('input.ts', createSimpleSchema());
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');

    fixture.source = `
      const query = \`
        query {
          hello
        }
      \`;
    `;
    expect(validateFn()).toEqual([]);
  });

  it('should return syntax error with empty template literal', () => {
    const fixture = craeteFixture('input.ts', createSimpleSchema());
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');

    fixture.source = `
      const query = \`\`;
    `;
    const actual = validateFn();
    expect(actual.length).toBe(1);
    expect((actual[0].messageText as string).match(/Syntax Error:/)).toBeTruthy();
  });

  it('should return position of error token', () => {
    const fixture = craeteFixture('input.ts', createSimpleSchema());
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');

    const frets: Frets = {};

    fixture.source = mark(
      // prettier-ignore
      '    const query = `query {`;    ' + '\n' +
      '%%%                      ^   %%%' + '\n' +
      '%%%                      a1  %%%' + '\n',
      frets,
    );
    const actual = validateFn();
    expect(actual[0].start).toBe(frets.a1.pos);
  });

  it('should return diagnostic array for invalid GraphQL template string', () => {
    const fixture = craeteFixture('input.ts', createSimpleSchema());
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');

    fixture.source = `
      const query = \`
        query {
          hoge
        }
      \`;
    `;
    const [actual] = validateFn();
    const messageText = actual.messageText as string;
    expect(/Cannot query field "hoge"/.test(messageText)).toBeTruthy();
  });

  it('should filter severity of GraphQL errors', () => {
    const fixture = craeteFixture('input.ts', createSimpleSchema());
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');

    fixture.source = `
      const query = \`
        query {
          helloWorld
        }
      \`;
    `;
    expect(validateFn()[0].category).toBe(ts.DiagnosticCategory.Warning);

    fixture.source = `
      const query = \`
        query {
          helo
        }
      \`;
    `;
    expect(validateFn()[0].category).toBe(ts.DiagnosticCategory.Error);
  });

  it('should work with incomplete document', () => {
    const fixture = craeteFixture('input.ts', createSimpleSchema());
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');

    fixture.source = `
      export const fragment = gql\`
        fragment MyFragment on Query {
          __typename

      \`;
    `;
    expect(validateFn().length).toBe(1);
  });

  it('should return "templateIsTooComplex" error when template node has too complex interpolation', () => {
    const fixture = craeteFixture('input.ts', createSimpleSchema());
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');

    const frets: Frets = {};
    fixture.source = mark(
      `
      const query = \`
        \${someFn()}
   %%%  \\ ^          %%%
   %%%  \\ a1         %%%
        query {
          ...MyFragment
        }
      \`;
    `,
      frets,
    );
    const actual = validateFn();
    expect(actual[0].code).toBe(ERROR_CODES.templateIsTooComplex.code);
    expect(actual[0].start).toBe(frets.a1.pos);
  });

  it('should return "isInOtherExpression" error when interpolated fragment has error', () => {
    const fixture = craeteFixture('input.ts', createSimpleSchema());
    const validateFn = fixture.adapter.getSemanticDiagnostics.bind(fixture.adapter, delegateFn, 'input.ts');

    const frets: Frets = {};
    fixture.source = mark(
      `
      const fragment = \`
        fragment MyFragment on Query {
          hoge
        }
      \`;
      const query = \`
        \${fragment}
   %%%  \\ ^          %%%
   %%%  \\ a1         %%%
        query {
          ...MyFragment
        }
      \`;
    `,
      frets,
    );
    const actual = validateFn();
    expect(actual.length).toBe(2);
    expect(actual[1].code).toBe(ERROR_CODES.errorInOtherInterpolation.code);
    expect(actual[1].start).toBe(frets.a1.pos);
  });
});
