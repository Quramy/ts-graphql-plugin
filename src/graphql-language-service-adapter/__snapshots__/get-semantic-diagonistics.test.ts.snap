// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSemanticDiagnostics should report errors coused by schema building 1`] = `"[ts-graphql-plugin] Schema build error: 'some schema build error' Check schema.graphql[1:1]"`;
